crankshaft/compiler.md

```diff
@@ -0,0 +1,281 @@
+
+- gather knowledge about types while program runs
+- **type dependent** code for operations given specific hidden classes as inputs
+- 1. validate type assumptions (are hidden classes as expected)
+- 2. do work
+- change at runtime via backpatching as more types are discovered to generate new ICs
+  [watch](http://youtu.be/UJPdhx5zTaw?t=28m44s) | [slide](http://v8-io12.appspot.com/index.html#56)
+
+Inline Caches alone without optimizing compiler step make huge performance difference (20x speedup).
+
+### Monomorphism vs. Polymorphism
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=31m30s) | [slide](http://v8-io12.appspot.com/index.html#61)
+
+- operations are monomorphic if hidden classes of arguments are **always** same
+- all others are polymorphic at best and megamorphic at worst
+- monomorphic operations are easier optimized
+
+#### Considerations
+
+- prefer monomorphic over polymorphic functions wherever possible
+
+### Optimizing Compiler
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=33m12s) | [slide](http://v8-io12.appspot.com/index.html#65)
+
+- if function executes a lot it becomes **hot**
+- hot function is re-compiled with optimizing compiler
+  - optimistically 
+  - lots of assumptions made from the calls made to that function so far
+  - type information takend from ICs
+  - operations get inlined speculatively using historic information
+  - monomorphic functions/constructors can be inlined entirely
+  - inlining allows even further optimizations
+
+### Deoptimization
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=36m50s) | [slide](http://v8-io12.appspot.com/index.html#78)
+
+- optimizations are speculative and assumptions are made
+- if assumption is violated
+  - function deoptimized
+  - execution resumes in full compiler code
+  - in short term execution slows down
+  - normal to occur
+  - more info about about function collected
+  - *better* optimization attempted
+  - if assumptions are violated again, deoptimized again and start over 
+- too many deoptimizations cause function to be sent to *deoptimization hell*
+  - considered not optimizable and no optimization is **ever** attempted again
+- certain constructs like `try/catch` are considered not optimizable and functions containing it go straight to
+  *deoptimization hell* due to **bailout** [watch](http://youtu.be/UJPdhx5zTaw?t=35m23s)
+
+None of this can be diagnosed with Chrome Devtools at this point.
+
+### Causes for Deoptimization
+
+#### Modifying Object Shape
+
+[watch](http://youtu.be/VhpdsjBUS3g?t=21m00s)
+
+- added fields (order matters) to object generate id of hidden class
+- adding more fields later on generates new class id which results in code using Point that now gets Point' to be
+  deoptimized
+
+[watch](http://youtu.be/VhpdsjBUS3g?t=21m45s)
+[watch](http://youtu.be/UJPdhx5zTaw?t=12m18s)
+
+```js
+function Point(x, y) {
+  this.x = x;
+  this.y = y;
+}
+
+var p = new Point(1, 2); // => hidden Point class created
```

Comment:
I agree that we should point that out (I'm not doing that in new doc either).
WRT an example I believe this one is more straightforward to understand, but I'll add a note WRT object literals.

Somehow github doesn't allow me to comment on previous comment.
I am aware of deopt cases that turned out to be v8 bugs, but the consensus is now to tell the V8 team about it instead of trying to workaround.

In the new compiler.md document I [do talk about deptimization](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#deoptimization), one [deopt loop bug](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#deoptimization-loop) that's been fixed and [an example](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#class-definitions-inside-functions) in which `usePoint` will essentially go to deoptimization hell.

If you have other examples we should add I can do that around that section.

Reply:
Addressed in 34a8f45f20f2d467f332a654ad10f17126e0feb9